### PR TITLE
feat: configurable logging style

### PR DESCRIPTION
**Pull request by NhanChau2409**
### Linked issue

Closes ducanhpham0312/zeevision-private#32

### Description
Logging out console under JSON format with zap for backend and tslog for frontend.
- In backend/internal/environment/environment.go, under init function, there is global logger object.
- In frontend/src/utils, there is logger.ts to initalize global logger

### Testing
Run the branch,
- For backend, check in terminal if errors, info, fatal have json format
- For frontend, at my best only snackbar.stories.tsx actually log out console, others are mainly examples of storybooks.

### CODE REVIEW

File: backend/cmd/zeevision/main.go [hunk 1]
```diff
@@ -31,12 +33,13 @@ func main() {
 	// Connect to database with retry
 	db, err := storage.ConnectDb(dsnConfig, DBConnectionRetries, DBConnectionRetryDelay)
 	if err != nil {
-		panic(err)
+		zap.L().Panic("Failed connecting to database", zap.Error(err))
 	}
+	zap.L().Info("Successfully connect to database")
```
Comment (kauppie) [inline backend/cmd/zeevision/main.go:38]: `connect` -> `connected`

File: backend/cmd/zeevision/main.go [hunk 2]
```diff
@@ -50,8 +53,10 @@ func main() {
 	brokers := []string{kafkaAddr}
 	kafkaConsumer, err := consumer.NewConsumer(storer, brokers, ConsumerRetries, ConsumerRetryDelay)
 	if err != nil {
-		panic(err)
+		zap.L().Panic("Failed connecting to Kafka", zap.Strings("brokers", brokers), zap.Error(err))
 	}
+	zap.L().Info("Successfully connect to Kafka", zap.Strings("brokers", brokers))
```
Comment (kauppie) [inline backend/cmd/zeevision/main.go:58]: `connect` -> `connected`

File: backend/cmd/zeevision/main.go [hunk 3]
```diff
@@ -83,16 +88,16 @@ func main() {
 		err = kafkaConsumer.ConsumeTopic(0, topic)
 		if err != nil {
 			// TODO: error handling
-			panic(err)
+			zap.L().Panic("Failed consuming", zap.String("topic", topic), zap.Error(err))
```
Comment (kauppie) [inline backend/cmd/zeevision/main.go:91]: You can remove the comment above

File: backend/cmd/zeevision/main.go [hunk 4]
```diff
@@ -83,16 +88,16 @@ func main() {
...
 		}
 	}
 
 	server, err := endpoint.NewFromEnv(fetcher)
 	if err != nil {
-		log.Fatal(err)
+		zap.L().Fatal("Failed creating a new endpoint from environment variables")
```
Comment (kauppie) [inline backend/cmd/zeevision/main.go:97]: You should print the causing error like you have done above with `zap.Error(err)`

File: backend/cmd/zeevision/main.go [hunk 5]
```diff
@@ -83,16 +88,16 @@ func main() {
...
 	}
 
 	if err := server.Run(); err != nil {
-		log.Fatal(err)
+		zap.L().Fatal("Errors happens when trying to run server")
```
Comment (kauppie) [inline backend/cmd/zeevision/main.go:101]: You can rephrase this to just `Server error`, since it will be printed for all errors during operation. Also include the causing error in the log.

File: backend/internal/consumer/kafka_consumer.go [hunk 6]
```diff
@@ -129,8 +129,11 @@ func (consumer *Consumer) ConsumeTopic(partition int32, topic string) (err error
 				// TODO: don't necessarily log every message like this
 				select {
 				case msgChannel <- msg.Value:
-					log.Printf("[%s/%d] Consumed message offset %d\n",
-						topic, partition, msg.Offset)
+					zap.L().Info("Consumed message",
+						zap.Int64("offset", msg.Offset),
+						zap.String("", topic),
+						zap.String("", "/"),
+						zap.Int32("", partition))
```
Comment (kauppie) [inline backend/internal/consumer/kafka_consumer.go:136]: Not sure what is happening here. Let `zap` do the formatting:

```golang
zap.L().Info("Consumed message",
    zap.Int64("offset", msg.Offset),
	zap.String("topic", topic),
	zap.Int32("partition", partition))
```

File: backend/internal/consumer/storage_updater.go [hunk 7]
```diff
@@ -21,7 +21,8 @@ type storageUpdater struct {
 	wg *sync.WaitGroup
 }
 
-func newDatabaseUpdater(storer storage.Storer, msgChannel msgChannelType, closeChannel signalChannelType, wg *sync.WaitGroup) *storageUpdater {
+func newDatabaseUpdater(storer storage.Storer, msgChannel msgChannelType, closeChannel signalChannelType,
+	wg *sync.WaitGroup) *storageUpdater {
```
Comment (kauppie) [inline backend/internal/consumer/storage_updater.go:25]: Seems like an unnecessary change, please revert

File: backend/internal/consumer/storage_updater.go [hunk 8]
```diff
@@ -58,13 +59,13 @@ readLoop:
 			var untypedRecord UntypedRecord
 			err := json.Unmarshal(msg, &untypedRecord)
 			if err != nil {
-				log.Printf("Failed to unmarshal: %v", err)
+				zap.L().Info("Failed to unmarshal: ", zap.Error(err))
```
Comment (kauppie) [inline backend/internal/consumer/storage_updater.go:62]: This is a failure, but it doesn't lead to program crashing. Change the logging level to `Error`.

File: backend/internal/consumer/storage_updater.go [hunk 9]
```diff
@@ -58,13 +59,13 @@ readLoop:
...
 				continue readLoop
 			}
 
 			err = u.handlingDispatch(&untypedRecord)
 			if err != nil {
-				log.Printf("Handling failed: %v", err)
+				zap.L().Info("Handling failed: ", zap.Error(err))
```
Comment (kauppie) [inline backend/internal/consumer/storage_updater.go:68]: Use `Error` here.

File: backend/internal/consumer/storage_updater.go [hunk 10]
```diff
@@ -103,8 +104,9 @@ func (u *storageUpdater) handlingDispatch(untypedRecord *UntypedRecord) error {
 			return fmt.Errorf("failed to handle incident: %w", err)
 		}
 	default:
-		log.Printf("Unhandled record: %v (intent: %v)",
-			untypedRecord.ValueType, untypedRecord.Intent)
+		zap.L().Info("Unhandled record:",
+			zap.String("value type", string(untypedRecord.ValueType)),
+			zap.String("intent", string(untypedRecord.Intent)))
```
Comment (kauppie) [inline backend/internal/consumer/storage_updater.go:109]: Use `Warn` here.

File: backend/internal/consumer/storage_updater.go [hunk 11]
```diff
@@ -189,11 +193,13 @@ func (u *storageUpdater) handleProcess(untypedRecord *UntypedRecord) error {
...
-			processDefinitionKey, bpmnProcessID)
+		zap.L().Info("Process created:",
+			zap.Int64("process definition", processDefinitionKey),
+			zap.String("bpmn process", bpmnProcessID))
 	default:
-		log.Printf("Unhandled intent for %v: %s",
-			record.ValueType, record.Intent)
+		zap.L().Info("Unhandled intent for ",
+			zap.String("value type", string(record.ValueType)),
+			zap.String("intent", string(record.Intent)))
```
Comment (kauppie) [inline backend/internal/consumer/storage_updater.go:202]: Use `Warn` here.

File: backend/internal/consumer/storage_updater.go [hunk 12]
```diff
@@ -234,31 +240,32 @@ func (u *storageUpdater) handleProcessInstance(untypedRecord *UntypedRecord) err
...
 				processInstanceKey,
 				time.UnixMilli(record.Timestamp),
 			)
 		}
 	default:
-		log.Printf("Unhandled intent for %v: %s",
-			record.ValueType, record.Intent)
+		zap.L().Info("Unhandled intent for ",
+			zap.String("value type", string(record.ValueType)),
+			zap.String("intent", string(record.Intent)))
```
Comment (kauppie) [inline backend/internal/consumer/storage_updater.go:268]: Use `Warn` here.

File: backend/internal/consumer/storage_updater.go [hunk 13]
```diff
@@ -279,26 +286,31 @@ func (u *storageUpdater) handleVariable(untypedRecord *UntypedRecord) error {
...
 			name,
 			value,
 			time.UnixMilli(record.Timestamp),
 		)
 	default:
-		log.Printf("Unhandled intent for %v: %s",
-			record.ValueType, record.Intent)
+		zap.L().Info("Unhandled intent for ",
+			zap.String("value type", string(record.ValueType)),
+			zap.String("intent", string(record.Intent)))
```
Comment (kauppie) [inline backend/internal/consumer/storage_updater.go:313]: Use `Warn` here.

File: backend/internal/environment/environment.go [hunk 14]
```diff
@@ -95,6 +97,9 @@ func init() {
 	setOrFallback(EnvVarDatabaseHost, DefaultHostDatabase)
 	setOrFallback(EnvVarDatabaseUser, "")
 	setOrFallback(EnvVarDatabasePassword, "")
+
+	// Create a global logger
+	zap.ReplaceGlobals(zap.Must(zap.NewProduction()))
```
Comment (kauppie) [inline backend/internal/environment/environment.go:102]: I would place this at the beginning of the `main` function since this file deals only with environment variables.

**Review by kauppie — CHANGES_REQUESTED** [review]
Initial reason for this issue was to include configurable logging options, so we can't really close the original issue without adding option to toggle between `json` and text formatted logs.

[Config](https://pkg.go.dev/go.uber.org/zap#Config) and its field `Encoding` fits the purpose for selecting the formatting style. Formatting style should be configurable with an environment variable.

In addition to adding the formatting style selection, please see the other comments.

File: backend/internal/consumer/storage_updater.go (hunk 7 above)
Comment (NhanChau2409) [reply backend/internal/consumer/storage_updater.go:25]: My gofmt lint auto did that. I try to put it back to norm but after save it auto lint again.

File: backend/internal/consumer/kafka_consumer.go [hunk 15]
```diff
@@ -129,8 +129,10 @@ func (consumer *Consumer) ConsumeTopic(partition int32, topic string) (err error
 				// TODO: don't necessarily log every message like this
 				select {
 				case msgChannel <- msg.Value:
-					log.Printf("[%s/%d] Consumed message offset %d\n",
-						topic, partition, msg.Offset)
+					zap.L().Info("Consumed message",
+						zap.Int64("offset", msg.Offset),
+						zap.String("", topic),
+						zap.Int32("", partition))
```
Comment (kauppie) [inline backend/internal/consumer/kafka_consumer.go:135]: Add `topic` and `partition` names instead of empty strings.

File: backend/internal/consumer/storage_updater.go (hunk 7 above)
Comment (kauppie) [reply backend/internal/consumer/storage_updater.go:25]: Try changing this to 
```golang
func newDatabaseUpdater(
	storer storage.Storer,
	msgChannel msgChannelType,
	closeChannel signalChannelType,
	wg *sync.WaitGroup,
) *storageUpdater {
```
so it looks clearer

File: backend/internal/consumer/storage_updater.go [hunk 16]
```diff
@@ -227,11 +233,11 @@ func (u *storageUpdater) handleProcessInstance(untypedRecord *UntypedRecord) err
 		timestamp,
 	)
 	if err != nil {
-		log.Printf(
+		zap.L().Error(
 			"Failed to log event to audit log: position %d, element id %s, err: %v",
```
Comment (kauppie) [inline backend/internal/consumer/storage_updater.go:237]: Change this to `Failed to log event to audit log` without `: position %d, element id %s, err: %v`

**Review by kauppie — CHANGES_REQUESTED** [review]
See the original change request I made. JSON/text formatting still isn't configurable.
